java/src/main/native/src/row_conversion.cu

```diff
@@ -0,0 +1,636 @@
+ */
+static std::unique_ptr<cudf::column> fixed_width_convert_to_rows(
+        const cudf::size_type start_row,
+        const cudf::size_type num_rows,
+        const cudf::size_type num_columns,
+        const cudf::size_type size_per_row,
+        std::unique_ptr<rmm::device_uvector<cudf::size_type>> & column_start,
+        std::unique_ptr<rmm::device_uvector<cudf::size_type>> & column_size,
+        std::unique_ptr<rmm::device_uvector<const int8_t *>> & input_data,
+        std::unique_ptr<rmm::device_uvector<const cudf::bitmask_type *>> & input_nm,
+        const cudf::scalar & zero,
+        const cudf::scalar & scalar_size_per_row,
+        cudaStream_t stream,
+        rmm::mr::device_memory_resource* mr
+        ) {
+    int64_t total_allocation = size_per_row * num_rows;
+    // We made a mistake in the split somehow
+    CUDF_EXPECTS(total_allocation < std::numeric_limits<int>::max(), "Table is too large to fit!");
+
+    // Allocate and set the offsets row for the byte array
+    std::unique_ptr<cudf::column> offsets = cudf::detail::sequence(
+            num_rows + 1, zero, scalar_size_per_row);
+
+    std::unique_ptr<cudf::column> data =
+        cudf::make_numeric_column(
+                cudf::data_type(cudf::type_id::INT8),
+                static_cast<cudf::size_type>(total_allocation));
+
+    dim3 blocks;
+    dim3 threads;
+    int shared_size = calc_fixed_width_kernel_dims(num_columns, num_rows, size_per_row, blocks, threads);
+
+    copy_from_fixed_width_columns<<<blocks, threads, shared_size, stream>>>(
+            start_row,
+            num_rows,
+            num_columns,
+            size_per_row,
+            column_start->data(),
+            column_size->data(),
+            input_data->data(),
+            input_nm->data(),
+            data->mutable_view().data<int8_t>());
+
+    return cudf::make_lists_column(num_rows,
+                std::move(offsets),
+                std::move(data),
+                0,
+                rmm::device_buffer{0, 0, mr});
+}
+
+static cudf::data_type get_data_type(const cudf::column_view & v) {
+    return v.type();
+}
+
+static bool is_fixed_width(const cudf::data_type & t) {
+    return cudf::is_fixed_width(t);
+}
+
+static inline int32_t align_offset(int32_t offset, std::size_t alignment) {
+    return (offset + alignment - 1) & ~(alignment - 1);
+}
+
+static inline bool are_all_fixed_width(std::vector<cudf::data_type> const & schema) {
+    return std::all_of(schema.begin(), schema.end(), cudf::java::is_fixed_width);
+}
+
+/**
+ * Given a set of fixed width columns, calculate how the data will be laid out in memory.
+ * @param [in] schema the types of columns that need to be laid out.
+ * @param [out] column_start the byte offset where each column starts in the row.
+ * @param [out] column_size the size in bytes of the data for each columns in the row.
+ * @return the size in bytes each row needs.
+ */
+static inline int64_t compute_fixed_width_layout(
```

Comment:
Nitpick: I have missed the reason that this returns an `int64_t`, when `at_offset` is 32bit, as is the return type for `align_offset()`.

Reply:
There is no reason, I should fix it.